src/Processors/Formats/Impl/DiagramOutputFrormat.cpp

```diff
@@ -0,0 +1,649 @@
+        }
+        writeString(String(diagram_symbols.bar) + "\n", out);
+    }
+    writeString(
+        y_min_str + diagram_symbols.left_bottom_corner + genRepeatString(diagram_symbols.dash, width / DiagramConstants::BRAILE_BLOCK_WIDTH)
+            + diagram_symbols.right_bottom_corner + '\n',
+        out);
+    if (width / DiagramConstants::BRAILE_BLOCK_WIDTH < x_min_str.size() + x_max_str.size())
+        writeString(String(DiagramConstants::FIELD_SIZE, ' ') + x_min_str + ' ' + x_max_str, out);
+    else
+        writeString(
+            String(DiagramConstants::FIELD_SIZE, ' ') + x_min_str
+                + String(width / DiagramConstants::BRAILE_BLOCK_WIDTH - x_min_str.size() - x_max_str.size(), ' ') + x_max_str,
+            out);
+}
+String DiagramOutputFormat::limitNumSize(Float64 num, size_t limit) const
+{
+    const int buffer_len = 100;
+    const int precision = 2;
+    char buffer[buffer_len];
+    int chars_written = std::snprintf(buffer, buffer_len, "%.*e", precision, num);
+    String str(buffer, chars_written);
+    if (chars_written < static_cast<int>(limit))
+        str = String(limit - 1 - static_cast<size_t>(chars_written), ' ') + str;
+    str += ' ';
+    return str;
+}
+
+void DiagramOutputFormat::writeSuffix()
+{
+    write();
+}
+std::pair<size_t, String>
+DiagramOutputFormat::getSerialzedStr(const IColumn & column, const ISerialization & serialization, size_t row_num, size_t max_size)
+{
+    String serialized_value = " ";
+    const DiagramSymbols & diagram_symbols = is_ascii_symbols ? ascii_diagram_symbols : utf_diagram_symbols;
+    WriteBufferFromString out_serialize(serialized_value, AppendModeTag());
+    serialization.serializeText(column, row_num, out_serialize, format_settings);
+    size_t serialized_value_visible_size = UTF8::computeWidth(reinterpret_cast<UInt8 *>(serialized_value.data()), serialized_value.size());
+    size_t new_size = serialized_value_visible_size;
+    if (serialized_value_visible_size > max_size)
+    {
+        serialized_value.resize(
+            UTF8::computeBytesBeforeWidth(reinterpret_cast<UInt8 *>(serialized_value.data()), serialized_value.size(), 0, max_size - 1));
+        serialized_value += diagram_symbols.ellipsis;
+        new_size = UTF8::computeWidth(reinterpret_cast<UInt8 *>(serialized_value.data()), serialized_value.size());
+    }
+    return {new_size, serialized_value};
+}
+ssize_t DiagramOutputFormat::BraileSymbol::getColor() const
+{
+    std::unordered_map<ssize_t, size_t> counts;
+    ssize_t most_common = -1;
+    size_t max_count = 0;
+    for (const auto & point_row : points)
+    {
+        for (const auto & point : point_row)
+        {
+            if (point != -1)
+            {
+                counts[point]++;
+                if (counts[point] > max_count)
+                {
+                    max_count = counts[point];
+                    most_common = point;
+                }
+            }
+        }
+    }
+    return most_common;
```

Comment:
Color tells which subpixel was the most common? Seems that heatmap coloring would be more useful.